Adds a couple of metrics computed on UMIs

fleharty opened the pull request:
### Description

This PR adds several metrics to UmiAwareMarkDuplicatesWithMateCigar.

These metrics will provide users concise statistics about the UMIs in their data.

The metrics added include:
UMI_LENGTH - the length of the UMIs in bases
OBSERVED_UNIQUE_UMIS - the number of observed UMIs in the deduplicated bam
INFERRED_UNIQUE_UMIS - the number of UMIs observed in the deduplicated bam after error correction
DUPLICATE_SETS_WITHOUT_UMI - the number of duplicate sets that would be identified without using UMIs
DUPLICATE_SETS_WITH_UMI - the number of duplicate sets identified by taking the UMIs into account
INFERRED_UMI_ENTROPY - entropy of inferred (error corrected) UMIs in base 4
OBSERVED_UMI_ENTROPY - entropy of all observed UMIs in base 4
UMI_BASE_QUALITIES - Phred scaled estimate of base qualities in UMIs.
UMI_COLLISION_EST - Estimate of the number of read pairs in the bam that are incorrectly labeled as belonging to the same duplicate set
UMI_COLLISION_Q - Phred scaled estimate of the probability a given read will have the same UMI as another read with the same start and stop position, but having come from a different original molecule

The reason for this PR is to provide users with metrics associated with UMIs.

----

### Checklist (never delete this)

Never delete this, it is our record that procedure was followed. If you find that for whatever reason one of the checklist points doesn't apply to your PR, you can leave it unchecked but please add an explanation below.

#### Content
- [x] Added or modified tests to cover changes and any new functionality
- [x] Edited the README / documentation (if applicable)
- [x] All tests passing on Travis

#### Review
- [x] Final thumbs-up from reviewer
- [x] Rebase, squash and reword as applicable

For more detailed guidelines, see https://github.com/broadinstitute/picard/wiki/Guidelines-for-pull-requests

yfarjoun commented:
please edit the description to the PR.

yfarjoun commented on src/main/java/picard/sam/markduplicates/UmiAwareMarkDuplicatesWithMateCigar.java:
```diff
@@ -65,6 +69,9 @@
     @Option(shortName = "MAX_EDIT_DISTANCE_TO_JOIN", doc = "Largest edit distance that UMIs must have in order to be considered as coming from distinct source molecules.", optional = true)
     public int MAX_EDIT_DISTANCE_TO_JOIN = 1;
 
+    @Option(shortName = "UMI_METRICS", doc = "UMI Metrics", optional = true)
```
the metrics_file should not be optional.

yfarjoun commented on src/main/java/picard/sam/markduplicates/UmiAwareMarkDuplicatesWithMateCigar.java:
```diff
@@ -78,13 +85,28 @@
+
+    @Override
+    protected int doWork() {
+        // Perform Mark Duplicates work
```
assert that metrics file is writable

yfarjoun commented:
back to you @fleharty

fleharty commented:
@yfarjoun Thanks for the comments, back to you.

yfarjoun commented on src/main/java/picard/sam/markduplicates/UmiAwareDuplicateSetIterator.java:
```diff
@@ -79,18 +96,21 @@
+        if (metrics.UMI_LENGTH > 0) {
+            collectMetrics();
+        }
+
```
extra newline

yfarjoun commented on src/main/java/picard/sam/markduplicates/UmiAwareDuplicateSetIterator.java:
```diff
@@ -119,6 +139,89 @@ private void process(final DuplicateSet set) {
+        for (DuplicateSet ds : duplicateSets) {
+            List<SAMRecord> records = ds.getRecords();
+            SAMRecord representativeRead = ds.getRepresentative();
+
```
extra newline

yfarjoun commented on src/test/java/picard/sam/markduplicates/UmiAwareMarkDuplicatesWithMateCigarTester.java:
```diff
@@ -155,6 +170,31 @@ public void test() {
             }
         }
+
+        if(expectedMetrics != null) {
```
need space between if and (

fleharty commented:
@yfarjoun Before you review, I need to fix the collectMetrics() function like we discussed on the phone.

fleharty commented:
@yfarjoun Back to you, thanks again for the review!

fleharty commented:
@yfarjoun back to you

fleharty commented:
@yfarjoun Made some minor changes to the javadoc.  Back to you.

yfarjoun commented on src/main/java/picard/sam/UmiMetrics.java:
```diff
@@ -24,48 +24,67 @@
+     * This estimate is computed over every duplicate set, and effectively accounts for the distribution of duplicate
+     * set sizes.  This is an experimental metric, and should be used with caution.
+     */
+    public double UMI_COLLISION_EST;
```
is this a rate? if so , call the metric `UMI_COLLISION_RATE`

yfarjoun commented on src/main/java/picard/sam/UmiMetrics.java:
```diff
@@ -24,48 +24,67 @@
 
     public void estimateBaseQualities(final int observedUmiBases) {
-        ESTIMATED_BASE_QUALITY_OF_UMIS = -10.0*Math.log10((double) OBSERVED_BASE_ERRORS / (double) observedUmiBases);
+        UMI_BASE_QUALITIES = -10.0*Math.log10((double) OBSERVED_BASE_ERRORS / (double) observedUmiBases);
```
there is a function QualUtil::getPhredScoreFromErrorProbability() which does this. cleaner and clearer to use that I think.

yfarjoun commented on src/main/java/picard/sam/UmiMetrics.java:
```diff
@@ -81,10 +100,67 @@ public UmiMetrics(final int length, final int observedUniqueUmis, final int infe
+        OBSERVED_UMI_ENTROPY = effectiveNumberOfBases(observedUmis);
+        INFERRED_UMI_ENTROPY = effectiveNumberOfBases(inferredUmis);
+
+        UMI_COLLISION_Q = -10 * Math.log10(UMI_COLLISION_EST / inferredUmis.size());
```
same here

yfarjoun commented on src/main/java/picard/sam/UmiMetrics.java:
```diff
@@ -81,10 +100,67 @@ public UmiMetrics(final int length, final int observedUniqueUmis, final int infe
+        for (Histogram.Bin observation : observations.values()) {
+            double pObservation = observation.getValue() / totalObservations;
+            entropyBase4 = entropyBase4 - pObservation * Math.log(pObservation);
+        }
```
can you replace this with 

```java
-observations.values().stream().collect(Collectors.summingDouble(v->v*Math.log(v)))/log(4.0);
```

yfarjoun commented on src/main/java/picard/sam/UmiMetrics.java:
```diff
@@ -81,10 +100,67 @@ public UmiMetrics(final int length, final int observedUniqueUmis, final int infe
+        return entropyBase4 / Math.log(4.0);
+    }
+
+    public void updateDuplicateSetMetrics(final int maxEditDistanceToJoin, final int duplicateSetSize) {
```
a public function needs a java-doc.. what is this function. what does it do? what are the inputs?

yfarjoun replied:
(or maybe it shouldn't be public...)

yfarjoun commented on src/main/java/picard/sam/UmiMetrics.java:
```diff
@@ -81,10 +100,67 @@ public UmiMetrics(final int length, final int observedUniqueUmis, final int infe
+
+        // For each duplicate set estimate the number of expected UMI collisions
+        double nWaysUmisCanDiffer = 0; // Number of ways two UMIs may contain errors, but be considered the same
+        for (int k = 0; k <= maxEditDistanceToJoin; k++) {
```
this doesn't seem to depend on duplicateSetSize. does it make sense to re-calculate it every time?

yfarjoun commented on src/main/java/picard/sam/UmiMetrics.java:
```diff
@@ -81,10 +100,67 @@ public UmiMetrics(final int length, final int observedUniqueUmis, final int infe
+        // pAllMembersCorrectlyLabeled = (1 - pCorrectlyLabeled)^duplicateSets.size().
+        // The expected number of reads incorrectly labeled as belonging to a duplicate set is given as,
+        // (1 - pAllMembersCorrectlyLabeled) * duplicateSets.size().
+        UMI_COLLISION_EST = UMI_COLLISION_EST +
```
why does the UMI_COLLISION_EST only grow?

yfarjoun replied:
I'm confused about this metric. can you explain it?

fleharty commented:
@yfarjoun Thanks again for the review.  Back to you.

I removed the UMI_COLLISION_EST and UMI_COLLISION_Q metrics since I don't think they did what I wanted them to do.

yfarjoun commented on src/main/java/picard/sam/UmiMetrics.java:
```diff
@@ -103,64 +87,27 @@ public UmiMetrics(final int length, final int observedUniqueUmis, final int infe
-        observedUmis.increment(currentUmi);
-        inferredUmis.increment(inferredUmi);
+        double entropyBase4 = observations.values().stream().collect(Collectors.summingDouble(
+                v -> -v.getValue() / totalObservations * Math.log(v.getValue() / totalObservations))) / Math.log(4.0);
```
the function itself could be made clearer:
```
v->{ double p = v.getValue() / totalObservations;
return -p*log(p)/log(4.0); 
}
```
Also, if you need log base 4 a lot, perhaps you should make a LOG_BASE_4 object in MathUtil and use that. it will save on recomputing log(4.0) every time and, again make things clearer

yfarjoun commented on src/main/java/picard/sam/markduplicates/UmiAwareDuplicateSetIterator.java:
```diff
@@ -151,11 +156,17 @@ private void process(final DuplicateSet set) {
                         }
                     }
-                    metrics.updateUmiRecordMetrics(currentUmi, inferredUmi);
+//                    metrics.updateUmiRecordMetrics(currentUmi, inferredUmi);
```
why the comment?

yfarjoun commented on src/main/java/picard/sam/UmiMetrics.java:
```diff
@@ -103,64 +87,27 @@ public UmiMetrics(final int length, final int observedUniqueUmis, final int infe
     }
 
-    public void calculateDerivedFields() {
+    public void calculateDerivedFields(final Histogram<String> observedUmis, Histogram<String> inferredUmis, long observedUmiBases) {
```
These should be private members in the metric...no need to pass them into a function.

yfarjoun commented on src/main/java/picard/sam/markduplicates/UmiAwareDuplicateSetIterator.java:
```diff
@@ -59,7 +62,9 @@
     private UmiMetrics metrics;
     private boolean haveWeSeenFirstRead = false;
-    private double expectedCollisions = 0;
+    private Histogram<String> observedUmis = new Histogram<>();
```
these should be part of the metric, not the iterator. ..unless you need them in the iterator.

yfarjoun commented on src/main/java/picard/sam/markduplicates/UmiAwareDuplicateSetIterator.java:
```diff
@@ -151,11 +156,17 @@ private void process(final DuplicateSet set) {
-        metrics.updateDuplicateSetMetrics(maxEditDistanceToJoin, duplicateSets.size());
+
+        metrics.DUPLICATE_SETS_WITH_UMI += duplicateSets.size();
+        metrics.DUPLICATE_SETS_WITHOUT_UMI++;
```
DUPLICATE_SETS_WITHOUT_UMI doesn't quite convey the meaning I think.

don't you mean DUPLICATE_SETS_IGNORING_UMI? or ORIGINAL_DUPLICATE_SETS ?

yfarjoun commented on src/test/java/picard/sam/markduplicates/UmiAwareMarkDuplicatesWithMateCigarTester.java:
```diff
@@ -191,8 +191,6 @@ public void test() {
             Assert.assertEquals(observedMetrics.OBSERVED_UMI_ENTROPY, expectedMetrics.OBSERVED_UMI_ENTROPY, tolerance, "OBSERVED_UMI_ENTROPY does not match expected");
             Assert.assertEquals(observedMetrics.UMI_BASE_QUALITIES, expectedMetrics.UMI_BASE_QUALITIES, tolerance, "UMI_BASE_QUALITIES does not match expected");
-            Assert.assertEquals(observedMetrics.UMI_COLLISION_EST, expectedMetrics.UMI_COLLISION_EST, tolerance, "UMI_COLLISION_EST does not match expected");
-            Assert.assertEquals(observedMetrics.UMI_COLLISION_Q, expectedMetrics.UMI_COLLISION_Q, tolerance, "UMI_COLLISION_Q does not match expected");
```
The UMI Collision sounds a bit like a Poission Binomial distribution....

fleharty commented:
@yfarjoun Thanks again for the review, back to you.

fleharty commented:
@yfarjoun I realized that I forgot to fix the lambda expression, so I did that and pushed it.

yfarjoun commented on src/main/java/picard/sam/markduplicates/UmiAwareDuplicateSetIterator.java:
```diff
@@ -156,17 +153,19 @@ private void process(final DuplicateSet set) {
                     observedUmiBases += currentUmi.length();
-                    observedUmis.increment(currentUmi);
-                    inferredUmis.increment(inferredUmi);
+                    metrics.observedUmis.increment(currentUmi);
```
Seeing how this is used, I would reccomend changing `observedUmis` and `inferredUmis` to private members of the metric and adding a function .addUmiObservation that takes both currentUmi and inferredUmi as input. this function should also update a private long `UmiBasesObserved` member to use in `calculateDerivedMetrics()`

yfarjoun commented on src/main/java/picard/sam/markduplicates/UmiMetrics.java:
```diff
@@ -82,14 +84,14 @@ public UmiMetrics(final int length, final int observedUniqueUmis, final int infe
     }
 
-    public void calculateDerivedFields(final Histogram<String> observedUmis, Histogram<String> inferredUmis, long observedUmiBases) {
+    public void calculateDerivedFields(long observedUmiBases) {
```
should not take any parameters.

fleharty commented:
@yfarjoun I made the histograms private, removed the argument from calculateDerivedFields.
I added a new public methods addUmiObservation to the metrics to increment the histograms.
Thanks again for the review.
Back to you @yfarjoun .

yfarjoun commented on src/main/java/picard/sam/markduplicates/UmiAwareMarkDuplicatesWithMateCigar.java:
```diff
@@ -78,13 +85,29 @@
+        IOUtil.assertFileIsWritable(UMI_METRICS_FILE);
+
+        // Perform Mark Duplicates work
+        int retval=super.doWork();
```
spaces around ` = `

yfarjoun commented on src/main/java/picard/sam/markduplicates/UmiMetrics.java:
```diff
@@ -0,0 +1,129 @@
+    private long observedUmiBases = 0;
+
+    /** Number of bases in each UMI */
+    public int UMI_LENGTH;
```
initialize

yfarjoun replied:
what if it's not the same? should it be an average?

yfarjoun commented on src/main/java/picard/sam/markduplicates/UmiMetrics.java:
```diff
@@ -0,0 +1,129 @@
+    public double INFERRED_UMI_ENTROPY = 0;
+
+    /** Estimation of Phred scaled quality scores for UMIs */
+    public double UMI_BASE_QUALITIES;
```
initialize

yfarjoun commented on src/test/java/picard/sam/markduplicates/UmiAwareMarkDuplicatesWithMateCigarTester.java:
```diff
@@ -155,6 +169,29 @@ public void test() {
+            Assert.assertEquals(observedMetrics.INFERRED_UNIQUE_UMIS, expectedMetrics.INFERRED_UNIQUE_UMIS, "INFERRED_UNIQUE_UMIS does not match expected");
+            Assert.assertEquals(observedMetrics.OBSERVED_BASE_ERRORS, expectedMetrics.OBSERVED_BASE_ERRORS, "OBSERVED_BASE_ERRORS does not match expected");
+            Assert.assertEquals(observedMetrics.DUPLICATE_SETS_IGNORING_UMI, expectedMetrics.DUPLICATE_SETS_IGNORING_UMI, "DUPLICATE_SETS_WITHOUT_UMI does not match expected");
+            Assert.assertEquals(observedMetrics.INFERRED_UMI_ENTROPY, expectedMetrics.INFERRED_UMI_ENTROPY, tolerance, "INFERRED_UMI_ENTROPY does not match expected");
```
please also check DUPLICATE_SETS_WITH_UMI

yfarjoun commented on src/main/java/picard/sam/markduplicates/UmiAwareDuplicateSetIterator.java:
```diff
@@ -79,18 +88,20 @@
         isOpen = false;
         wrappedIterator.close();
+
+        if (metrics.UMI_LENGTH > 0) {
```
I don't understand why this if statement...what will happen if the UMI_LENGTH is 0?

yfarjoun commented on src/main/java/picard/sam/markduplicates/UmiAwareDuplicateSetIterator.java:
```diff
@@ -119,6 +130,42 @@ private void process(final DuplicateSet set) {
+                        haveWeSeenFirstRead = true;
+                    } else {
+                        if (metrics.UMI_LENGTH != currentUmi.length()) {
+                            throw new PicardException("UMIs of differing lengths were found.");
```
Should there be a TODO to allow UMI's of different length? perhaps UMI_LENGTH should be a double to mean "average" umi length even if we do not allow that right now?

fleharty commented:
@yfarjoun if you have a moment, could you look at this again?  Thanks!

yfarjoun commented:
I did..I'm not seeing an additional commit from you..am I missing something?

fleharty commented:
@yfarjoun my apologies, somehow I didn't see your most recent comments... I'll have them fixed soon

fleharty commented:
@yfarjoun Thanks again for the review.  Back to you.
